src/BatchProcessor.ts

```diff
@@ -1,19 +1,245 @@
+function humanify(when?: number, flags: {
+  parts?: boolean
+  terse?: boolean
+} = {}) {
+  const d = when ? new Date(when) : new Date()
+  const iso = d.toISOString()
+
+  if (flags.parts) {
+    let parts = iso.split(/[-:T.Z]/).map(s => +s)
+    let i = 0
+    let out: any = {
+      year: parts[i++],
+      month: parts[i++],
+      day: parts[i++],
+      hour: parts[i++],
+      minute: parts[i++],
+      second: parts[i++],
+      milli: parts[i++],
+    }
+    if (flags.terse) {
+      out = {
+        ty: out.year,
+        tm: out.month,
+        td: out.day,
+        th: out.hour,
+        tn: out.minute,
+        ts: out.second,
+        ti: out.milli,
+      }
+    }
+    return out
+  }
+
+  return +(iso.replace(/[^\d]/g, '').replace(/\d$/, ''))
+}
 
 function BatchProcessor(this: any, options: BatchProcessorOptionsFull) {
   const seneca: any = this
+  
+  
+  const Patrun = seneca.util.Patrun
+  const Jsonic = seneca.util.Jsonic
+  const Deep = seneca.util.deepextend
+  const generate_id = seneca.util.Nid
+  const BatchId = generate_id()
+  
+  let wheres: any = new Patrun({ gex: true })
+  
+  for(const message_whence in options.where) {
+    let match = options.where[message_whence]?.match
+    let pattern = Jsonic(message_whence)
+    for(const message_pattern in match) {
+      let workflow = match[message_pattern]
+      // console.log(message_pattern, pattern)
+      if(ALL == message_pattern) {
+        let matchInst = wheres.find(pattern) || new Match(new Patrun({ gex: true }))
+        matchInst.set_all(workflow)
+        wheres.add(pattern, matchInst)
+      } else {
+        let matchInst = wheres.find(pattern) || new Match(new Patrun({ gex: true }))
+        matchInst.add_pattern(Jsonic(message_pattern), workflow)
+        wheres.add(pattern, matchInst)
+        
+      }
+    
+    }
+  }
 
-  function process(seneca: any, ctx: any, out: any) {
+  async function process(this: any, seneca: any, ctx: any, out: any = {}) {
+  
+    const BatchMonitorEntry = ctx.BatchMonitorEntry$ || function(this: any, ...args: any) {}
+    const whence = seneca.private$?.act?.msg
+    
+    // console.log(whence, wheres.list())
+    let where = wheres.find(whence)
+    let workflow: any = null
+    let results = ctx.result$ = ctx.result$ || []
+    
+    
+    // console.log(seneca.private$?.act?.msg)
+    
+    out = { ...out }
+
+    out.run = out.run || ('R' + generate_id())
+    out.batch = out.batch || ('B' + humanify())
+    
+    if(null == where) {
+      throw new Error("whence not found!")
+    }
+    
+    
+    if(workflow = where.find(out)) {
+      // let i = 0;
+      const send = Array.isArray(workflow.send) ? workflow.send : [ workflow.send ]
+      for(let config of send) {
+        let { msg, body } = config
+        let msg_evld = Utility.evaluateMessage(seneca, ctx, out, msg, body)
+        
+        // console.log(config, msg_evld)
+        
+        Utility.workflowRun(seneca, msg_evld, config, options, results)
```

Comment:
workflowRun is not a utility, it is core logic, it just needs to be a separate function